subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/ProviderBackedFileCollection.java

```diff
@@ -56,7 +56,10 @@ public void visitDependencies(TaskDependencyResolveContext context) {
     @Override
     protected void visitChildren(Consumer<FileCollectionInternal> visitor) {
         UnpackingVisitor unpackingVisitor = new UnpackingVisitor(visitor, resolver, patternSetFactory);
-        unpackingVisitor.add(provider.get());
+        Object providerValue = provider.getOrNull();
+        if (providerValue != null) {
+            unpackingVisitor.add(providerValue);
+        }
```

Comment:
You can pass `null` to `UnpackingVisitor`, so this `null`-check is unnecessary.

Reply:
Another option would be to add `Provider.ifPresent(Consumer<T>)` and use it here.